WIP util/mon: more reduction in size

yuzefovich opened the pull request:
Epic: None

yuzefovich committed:
util/mon: remove another field

This commit removes `configLimit` field from the monitor struct that
stores the limit that was specified when the monitor was created and was
introduced in 81c8983a703463542bb857135fd7d8f998ea5e15.

This commit refactors the logic to achieve what that change wanted by
using the current value of `limit` as the `configLimit`. This provides
exactly the same behavior _assuming_ that the same monitor is not
started after having been stopped (which doesn't happen in the use
case the linked commit had in mind - inheriting from the temp storage
config monitor).

Generally, this pattern (of stopping and starting the same monitor) is
not common. One example where we use this is in `resetForNewSQLTxn`, but
there we don't need this "inheritance across restarts" property AFAICT.

A couple of tests needed an adjustment to remove this pattern.

Release note: None

yuzefovich committed:
WIP on asserting that parent has been started

yuzefovich committed:
util/mon: remove settings object from the monitor

This allows us to reduce the size of the struct into size class 144
bytes. The downside is that we won't have the settings to see whether
diagnostics are enabled so that a sentry report is sent if we're trying
to shrink the account by more than it has. In other words, we will no
longer receive "unexpected leftover bytes" nor "no bytes to release"
reports (which were usually unactionable).

Note that `ReportOrPanic` will still panic in test builds hopefully
providing good enough coverage of misuse of the accounting system.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120534)
<!-- Reviewable:end -->

yuzefovich committed:
util/mon: store the pointer to IHistogram interface

This brings down the size of the monitor struct by another 8 bytes. This
approach seems acceptable given that only a handful of monitors specify
non-nil MaxHist option, so the amount of extra pointers floating around
should be relatively small.

Release note: None

yuzefovich commented:
Closing since I don't think it's that important to reduce the size further.
